[dune] Add "quick" and "check" targets for fast builds.

ejgallego opened the pull request:
In #8900 a quicker build target is requested, this PR does provide
targets towards that goal.

- `check`, introduced in Dune 1.5.0, will build all ml files in a fast way,
- `quick`, does build a set of hand-picked targets in a fast way.

Further improvements could come from having `coq_dune` use
`coqtop.byte` when appropriate, taking advantage from
https://github.com/ocaml/dune/issues/1073, and not generating rules
for `.vo` files.

gares commented:
> `quick`, does build a set of hand-picked targets in a fast way.

Can you explain which targets are built and why this set is relevant? Which is the rationale/use-case?
What can I do with these things (re)built? Can I run ocamldebug? Can I load plugins?

ejgallego commented:
> Can you explain which targets are built and why this set is relevant? Which is the rationale/use-case?

These targets are mainly derived from what it is already on the Makefiles in different forms. For example `pluginsopts`, etc... I added them because some developers requested it, I don't use them myself. AFAIU they compile most of the ML part of Coq.

> What can I do with these things (re)built?

Run them?

> Can I run ocamldebug? Can I load plugins?

Yup, that should work. Tho for ocamldebug you have the `dune-dbg` target that should work better. In the future Dune will support running ocamldebug directly I think.

ejgallego commented:
> AFAICT the docker update is unecessary (as CI doesn't test the new targets) and puts us out of sync with our declared dependencies (eg in .opam files).

Yup that's a problem but hopefully soon Dune will generate the opam files and take care of adding the right deps there.

> I think you should remove the docker changes and put "requires dune >= 1.5.0" in the make help for check.
> You could also bump our required dune version throughout, although I'm not sure it's worth doing here since only the new feature will break without updated dune.

I don't mind either way, Dune is released quite often so updates are expected.

gares commented:
> These targets are mainly derived from what it is already on the Makefiles in different forms. For example `pluginsopts`

pluginsopt builds native code, here I see `.bc` stuff.
Unfortunately, for my workflow, byte code is unusable (and ocamldebug too).

In any case I'm fine with a target for a minimal bytecode build. But you should find a name that says so. quick is a bit too vague, IMO. For sure it should mention "bytecode".

ejgallego commented:
> I don't mind either way, Dune is released quite often so updates are expected.

I have removed the 1.5.0 requirement.

> pluginsopt builds native code, here I see `.bc` stuff.
> Unfortunately, for my workflow, byte code is unusable (and ocamldebug too).
> 
> In any case I'm fine with a target for a minimal bytecode build. But you should find a name that says so. quick is a bit too vague, IMO. For sure it should mention "bytecode".

Oh indeed, I've added a `quickbyte` and `quickopt` then.

gares commented on Makefile.dune:
```diff
@@ -34,6 +37,21 @@ world: voboot
+PLUGIN_FILES=$(wildcard plugins/*/*.mlpack)
+PRINTER_FILES=dev/top_printers.cma dev/checker_printers.cma
+QUICKBYTE_TARGETS=$(COQTOP_FILES) $(PLUGIN_FILES:.mlpack=.cma) $(PRINTER_FILES) topbin/coqtop_byte_bin.bc
+QUICKOPT_TARGETS=$(COQTOP_FILES:.bc=.exe) $(PLUGIN_FILES:.mlpack=.cmxa) $(PRINTER_FILES:.cma=.cmxa) topbin/coqtop_bin.exe
```
I'm not so sure that PRINTER_FILES in opt are of any use. They are typically loaded the the ocaml toplevel for getting better `#trace` display.

ejgallego replied:
They are included in `ocamlopt` which is what this target emulates. Maybe I should call it this way? Anyways, it is important to compile them and they take very little time.

Zimmi48 commented on Makefile.dune:
```diff
@@ -1,7 +1,7 @@
 # Dune Makefile for Coq
 
-.PHONY: help voboot states world watch test-suite release apidoc ocheck ireport clean
+.PHONY: help voboot states world watch check quick test-suite release apidoc ocheck ireport clean
```
This has gone out of sync with the rest of the patch.

SkySkimmer commented:
@Zimmi48 a bit of make won't kill you, but whatever I'll do it.

ejgallego commented:
> I thought that by agreeing to be secondary code owner for dune files, it would prevent me having to read any Makefile-syntax and already this is wrong.

This is a very good point, I could implement the rules using Dune syntax, but I didn't do so as I thought this would be more familiar to devs.

SkySkimmer commented:
>This is a very good point, I could implement the rules using Dune syntax, but I didn't do so as I thought this would be more familiar to devs.

I think implementing them in dune syntax would help devs learn from example how to write their own rules.

ejgallego commented:
> I think implementing them in dune syntax would help devs learn from example how to write their own rules.

A bit of the idea of Dune is that devs should not have to write any rules, so I am not sure we want to add a gazillion Dune targets, but indeed, I'll give it a try.

ejgallego commented:
> A bit of the idea of Dune is that devs should not have to write any rules, so I am not sure we want to add a gazillion Dune targets, but indeed, I'll give it a try.

Well, basically the Dune rule would list all the targets, so it is pretty boring:
```
(alias
  (name plugins)
  (deps plugin1 ... pluginn))
```
There is no mechanism to generate stuff like is done in make, and indeed there are technical reasons for that. When some dynamic rules are needed usually `include` is used with `promote`.

So I am not sure what to here at this stage of the Dune implementation. Maybe make would be the most convenient solution for now.

SkySkimmer commented:
OK then fix the PHONY then I'll merge.

ejgallego commented:
Fixed thanks!
